File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/block/BlockDoTest.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.mChain;
import static com.flat.simplex.support.TokenChain.readChain;
import static org.junit.jupiter.api.Assertions.*;

class BlockDoTest {

    @Test
    public void block() {
        TokenChain chain = readChain("do hello();");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();

        assertTrue(block.isLoop());
        assertFalse(block.isCommandBlock());
        TokenChain.assertChain("hello();", block.getTokenContent(), block.getTokenContentEnd(), "Invalid body");

        assertErrors(context);
    }

    @Test
        public void blockDoBlock() {
        TokenChain chain = readChain("do{}");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();

        assertTrue(block.isCommandBlock());
        TokenChain.assertOne("{}", block.getTokenContent(), "Invalid body");

        assertErrors(context);
    }

    @Test
    public void blockDoMarkWhile() {
        TokenChain chain = readChain("do hello();");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();

        BlockWhile blockWhile = getBlockWhile(context);
        block.markWhile(blockWhile);

        assertFalse(block.isCommandBlock());
        assertEquals(blockWhile, block.getBlockWhile(), "Incorrect block While");
        TokenChain.assertChain("hello();", block.getTokenContent(), block.getTokenContentEnd(), "Invalid body");

        assertErrors(context);
    }

    @Test
    public void blockDoMarkWhileBlock_Fail() {
        TokenChain chain = readChain("do hello();");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();

        BlockWhile blockWhile = getBlockWhileBlock(context);
        block.markWhile(blockWhile);

        assertFalse(block.isCommandBlock());
        TokenChain.assertChain("hello();", block.getTokenContent(), block.getTokenContentEnd(), "Invalid body");

        assertErrors(context, Error.doWhileUnexpectedBlock);
    }

    @Test
    public void blockDoMarkNoWhile_Fail() {
        TokenChain chain = readChain("do hello();");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();
        block.markWhile(null);

        assertFalse(block.isCommandBlock());
        TokenChain.assertChain("hello();", block.getTokenContent(), block.getTokenContentEnd(), "Invalid body");

        assertErrors(context, Error.doWhileExpected);
    }

    @Test
    public void blockDoUnexpectedToken_Fail() {
        TokenChain chain = readChain("do {};");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();

        assertTrue(block.isCommandBlock());
        TokenChain.assertOne("{}", block.getTokenContent(), "Invalid body");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void blockDoUnexpectedEndOfTokens_Fail() {
        TokenChain chain = readChain("do");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();

        assertFalse(block.isCommandBlock());
        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void blockDoOpenBlock_Fail() {
        TokenChain chain = readChain("do{");

        Context context = new Context();
        BlockDo block = new BlockDo(context, null, chain.get(), null);
        block.read();

        assertTrue(block.isCommandBlock());
        TokenChain.assertOne("{", block.getTokenContent(), "Invalid body");

        assertErrors(context, Error.missingCloser);
    }

    private BlockWhile getBlockWhile(Context context) {
        TokenChain chain = readChain("while(true);");

        return new BlockWhile(context, null, chain.get(), null);
    }

    private BlockWhile getBlockWhileBlock(Context context) {
        TokenChain chain = readChain("while(true){}");

        return new BlockWhile(context, null, chain.get(), null);
    }
}

File: SimplexApp/simplex/lang/parser/logic/block/BlockDo.test.js
const TokenChain = require("simplex/lang/support/TokenChain.js");
const Context = require("simplex/lang/parser/logic/Context.js");
const Error = require("simplex/lang/parser/logic/error/Error.js");
const BlockDo = require("simplex/lang/parser/logic/block/BlockDo.js");
const BlockWhile = require("simplex/lang/parser/logic/block/BlockWhile.js");
const ContextSupport = require("simplex/lang/support/ContextSupport.js");
const AssertError = require("simplex/lang/support/AssertError.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;
const assertTrue = AssertError.assertTrue;
const assertFalse = AssertError.assertFalse;
const assertEquals = AssertError.assertEquals;

test('block', () => {
    let chain = readChain('do hello();');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    assertTrue(block.isLoop());
    assertFalse(block.isCommandBlock());
    TokenChain.assertChain('hello();', block.getTokenContent(), block.getTokenContentEnd(), 'Invalid body');
    assertErrors(context);
});
test('blockDoBlock', () => {
    let chain = readChain('do{}');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    assertTrue(block.isCommandBlock());
    TokenChain.assertOne('{}', block.getTokenContent(), 'Invalid body');
    assertErrors(context);
});
test('blockDoMarkWhile', () => {
    let chain = readChain('do hello();');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    let blockWhile = getBlockWhile(context);
    block.markWhile(blockWhile);
    assertFalse(block.isCommandBlock());
    assertEquals(blockWhile, block.getBlockWhile(), 'Incorrect block While');
    TokenChain.assertChain('hello();', block.getTokenContent(), block.getTokenContentEnd(), 'Invalid body');
    assertErrors(context);
});
test('blockDoMarkWhileBlock_Fail', () => {
    let chain = readChain('do hello();');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    let blockWhile = getBlockWhileBlock(context);
    block.markWhile(blockWhile);
    assertFalse(block.isCommandBlock());
    TokenChain.assertChain('hello();', block.getTokenContent(), block.getTokenContentEnd(), 'Invalid body');
    assertErrors(context, Error.doWhileUnexpectedBlock);
});
test('blockDoMarkNoWhile_Fail', () => {
    let chain = readChain('do hello();');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    block.markWhile(null);
    assertFalse(block.isCommandBlock());
    TokenChain.assertChain('hello();', block.getTokenContent(), block.getTokenContentEnd(), 'Invalid body');
    assertErrors(context, Error.doWhileExpected);
});
test('blockDoUnexpectedToken_Fail', () => {
    let chain = readChain('do {};');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    assertTrue(block.isCommandBlock());
    TokenChain.assertOne('{}', block.getTokenContent(), 'Invalid body');
    assertErrors(context, Error.unexpectedToken);
});
test('blockDoUnexpectedEndOfTokens_Fail', () => {
    let chain = readChain('do');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    assertFalse(block.isCommandBlock());
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('blockDoOpenBlock_Fail', () => {
    let chain = readChain('do{');
    let context = new Context(chain.get());
    let block = new BlockDo(context, chain.get(), null);
    block.read();
    assertTrue(block.isCommandBlock());
    TokenChain.assertOne('{', block.getTokenContent(), 'Invalid body');
    assertErrors(context, Error.missingCloser);
});
function getBlockWhile(context) {
    let chain = readChain('while(true);');
    return new BlockWhile(context, chain.get(), null);
}
function getBlockWhileBlock(context) {
    let chain = readChain('while(true){}');
    return new BlockWhile(context, chain.get(), null);
}
